src/dune_engine/fs_notify_memo.ml

```diff
@@ -8,7 +9,20 @@ let memo =
     ~visibility:Hidden
     (fun _path -> Memo.Build.return ())
 
-let depend path = Memo.exec memo path
+(* Currently, we do not expose this low-level primitive. If you need it, perhaps
+   you could add a higher-level primitive instead, such as [file_exists]? *)
+let depend path =
+  if Path.is_in_build_dir path then
+    Code_error.raise "Fs_notify_memo.depend called on a build path" [];
+  Memo.exec memo path
+
+let file_exists path =
+  let+ () = depend path in
+  Path.exists path
+
+let file_digest path =
+  let+ () = depend path in
+  Cached_digest.source_or_external_file path
```

Comment:
I see you're "demoting" `Cached_digest` into a level of abstraction "below" file watching. I haven't thought too much about which direction makes more sense, but if you can share your thoughts it would be cool.

Reply:
My thinking was that computing the digest is similar to other `Path.t -> 'a` functions: reading the contents of a file, checking if a file exists, etc. The only unusual thing about it is that it's cached, but that seems more like an implementation detail. The `Fs_notify_memo` calls out to these functions and adds file watching on top, so it seems higher-level from that point of view.

Reply:
Hum, to me "being cached" seems like a thing that should ideally interact with fs notifications (be invalidated when notification arrives), so I don't see it as simply an implementation detail. Maybe when you say "at the bottom as well" you mean we can take care of that, but it's something that we'll have to do.

Reply:
You convinced me that there is something to think about here, so I added the following comment in my upcoming PR with directory tracking:

```ocaml
(* CR-someday amokhov: It is unclear if we got the layers of abstraction right
   here. One could argue that caching is a higher-level concept compared to file
   watching, and we should expose this function from the [Cached_digest] module
   instead. For now, we keep it here because it seems nice to group all tracked
   file system access functions in one place, and exposing an uncached version
   of [file_digest] seems error-prone. We may need to rethink this decision. *)
let file_digest = ...
```